### ISSUE TITLE
Opening SRTIS output as `AnalysisStorage` takes forever

### ISSUE THREAD

**Issue by dwhswenson** [openpathsampling/openpathsampling#559]
This may be a general problem related to the fact that SRTIS creates a lot of steps from dummy movers, and therefore this is a problem in general for systems with lots of steps. However, opening a 1.2GB file for analysis took over an hour and a half, whereas opening it as a regular readable storage took about a minute (which is already pretty long, in my opinion).

From `ncdump -h` on the file:

```
dimensions:
    scalar = 1 ;
    pair = 2 ;
    stores = UNLIMITED ; // (25 currently)
    trajectories = UNLIMITED ; // (15980 currently)
    topologies = UNLIMITED ; // (1 currently)
    cvs = UNLIMITED ; // (3 currently)
    snapshots = UNLIMITED ; // (554848 currently)
    snapshottype = UNLIMITED ; // (1 currently)
    cvcache = UNLIMITED ; // (3 currently)
    samples = UNLIMITED ; // (26381 currently)
    samplesets = UNLIMITED ; // (246001 currently)
    movechanges = UNLIMITED ; // (998173 currently)
    steps = UNLIMITED ; // (246001 currently)
    details = UNLIMITED ; // (796808 currently)
    pathmovers = UNLIMITED ; // (110 currently)
    shootingpointselectors = UNLIMITED ; // (8 currently)
    engines = UNLIMITED ; // (1 currently)
    pathsimulators = UNLIMITED ; // (1 currently)
    transitions = UNLIMITED ; // (3 currently)
    networks = UNLIMITED ; // (1 currently)
    schemes = UNLIMITED ; // (1 currently)
    interfacesets = UNLIMITED ; // (3 currently)
    msouters = UNLIMITED ; // (1 currently)
    volumes = UNLIMITED ; // (45 currently)
    ensembles = UNLIMITED ; // (155 currently)
    tag = UNLIMITED ; // (0 currently)
    snapshot0 = UNLIMITED ; // (554848 currently)
    snapshot0n_atoms = 1 ;
    snapshot0n_spatial = 2 ;
```

I wrote out a smaller SRTIS file and ran performance analysis on opening it for analysis. I've attached the pstats file in a zip. I've also included a visualization of the call graph from that.

[srtis_analysis.pstats.zip](https://github.com/openpathsampling/openpathsampling/files/415127/srtis_analysis.pstats.zip)

![srtis_analysis](https://cloud.githubusercontent.com/assets/8178546/17620101/286892be-608a-11e6-8ba6-599095a21078.png)

**Comment by jhprinz** [comment]
Is it possible that you exceed the standard cache sizes? I am not sure if the initial cache_all will be aware of too small caches. This will result in filling the cache once and then start loading items and removing others one by one and it could be that this will increase loading times A LOT.

First. Can you look at the logger and tell me which caching took longest? There should be at INFO level.

Could you then try to load everything with maximum cache? You can do so by
```py
st = paths.Storage('name', 'r')
# set all caches to max and hope for a big memory
for name, store in st._stores.items():
    store.set_caching(True)
# and run the cachings one by one
paths.AnalysisStorage.cache_for_analysis(st)
```

Try if that speeds it up (hopefully it should a lot).

The reason for opening it normally taking long is that CVs are always cached (I think). This is to ensure being complete.

**Comment by dwhswenson** [comment]
> Can you look at the logger and tell me which caching took longest?

Not for another 1.5 hours ;-) Running now, with output so far of:

```
(INFO)openpathsampling.netcdfplus.netcdfplus: Open existing netCDF file 'srtis.nc' for reading - reading from existing file
(INFO)openpathsampling.storage.storage: Cached all CVs in 29382 ms
(INFO)openpathsampling.storage.storage: Cache all objects of store `cvs` [3] in 19 ms
(INFO)openpathsampling.storage.storage: Cache all objects of store `samples` [26381] in 227245 ms
(INFO)openpathsampling.storage.storage: Cache all objects of store `samplesets` [246001] in 156528 ms
(INFO)openpathsampling.storage.storage: Cache all objects of store `volumes` [45] in 38 ms
(INFO)openpathsampling.storage.storage: Cache all objects of store `ensembles` [155] in 9 ms
(INFO)openpathsampling.storage.storage: Cache all objects of store `pathmovers` [110] in 980 ms
```

> Could you then try to load everything with maximum cache?

I get an error. Running (with blank lines representing cell breaks in an ipynb):

```python
import openpathsampling as paths

%time storage = paths.Storage(filename="srtis.nc")

import logging.config
logging.config.fileConfig("test_logging.conf", disable_existing_loggers=False)

%%time
for name, store in storage._stores.items():
    store.set_caching(True)

%%time
paths.AnalysisStorage.cache_for_analysis(storage)
```

I get the output:

```pytb
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-fe97d98c94e0> in <module>()
----> 1 get_ipython().run_cell_magic(u'time', u'', u'paths.AnalysisStorage.cache_for_analysis(storage)')

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc in run_cell_magic(self, magic_name, line, cell)
   2101             magic_arg_s = self.var_expand(line, stack_depth)
   2102             with self.builtin_trap:
-> 2103                 result = fn(magic_arg_s, cell)
   2104             return result
   2105 

<decorator-gen-59> in time(self, line, cell, local_ns)

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/IPython/core/magic.pyc in <lambda>(f, *a, **k)
    186     # but it's overkill for just that one bit of state.
    187     def magic_deco(arg):
--> 188         call = lambda f, *a, **k: f(*a, **k)
    189 
    190         if callable(arg):

/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/IPython/core/magics/execution.pyc in time(self, line, cell, local_ns)
   1170         if mode=='eval':
   1171             st = clock2()
-> 1172             out = eval(code, glob, local_ns)
   1173             end = clock2()
   1174         else:

<timed eval> in <module>()

/Users/dwhs/Dropbox/msm-tis/openpathsampling/storage/storage.pyc in cache_for_analysis(storage)
    457             for cv, (cv_store, cv_store_idx) in \
    458                     storage.snapshots.cv_list.items():
--> 459                 cv_store.cache.load_max()
    460 
    461         stores_to_cache = ['cvs',

AttributeError: 'MaxCache' object has no attribute 'load_max'
```

**Comment by jhprinz** [comment]
I know what happens. It is doing what I described but only for movechanges. 

I will fix this. Can you try setting only movechanges to max with

    st.set_caching_mode('analysis')
    st.movechanges.set_caching(True)

**Comment by dwhswenson** [comment]
And then do `paths.AnalysisStorage.cache_for_analysis(st)`?

**Comment by dwhswenson** [comment]
Better, but still pretty rough. Over 30 minutes.

Before `storage.movechanges.set_caching(True)`:
```
(INFO)openpathsampling.storage.storage: Cache all objects of store `movechanges` [998173] in 3934924 ms
```

After:

```
(INFO)openpathsampling.storage.storage: Cache all objects of store `movechanges` [998173] in 1425484 ms
```

So that is sped up by a factor of ~2.75 -- but that's still a really long time to open a file that is only 1.2 GB.

**Comment by jhprinz** [comment]
Okay these are 1 Million objects, that need to be loaded bz string unpickled and created. Including looking up all the connected objects. I think this will be slow unless we start implementing it using cython.

What we can do is to not load all the steps in advance. Meaning that we only load the steps, but not the complete pathmovechanges associated with it. This should make it much faster and all of the rate computation stuff should work. All the specific move analysis would be slow though...

**Comment by dwhswenson** [comment]
Next step in the analysis in SRTIS notebook is calculating acceptance of various movers (and this should always be done to check for red flags). I think this requires the `Changes`. (It also takes an agonizingly long time -- 45 minutes back when loading took 1h30. I was going to profile that next.)

I think that Cythonization of storage should be a relatively high 1.x priority. Storage is much of our overhead, and I don't know if there are any optimizations short of Cythonization that will help.

**Comment by jhprinz** [comment]
I think this will affect toy simulations mostly. Having 250000 steps even in an alanine simulation with more then 500k snapshots would already be a 50GB file, correct? This seems a lot for Alanine. And then spending 45 minutes to analyze is not super bad, considering how it takes to generate the data.

I understand that this is bad for toy systems but (for now) this is the price to pay for the storage convenience we have. 

Question is, how it will take to fix this and what would be an acceptable time. If you run a bigger system it will be slow again at some point.

**Comment by dwhswenson** [comment]
With the current setup of SRTIS, this is likely to be a problem with SRTIS for real systems, too. The dummy moves are a killer. (See my constant ranting about how ugly this approach is.)

**Comment by jhprinz** [comment]
That is true, but will that we solvable? I am just too inexperience on how easy it is to replace some of the loading / saving of objects with cython.

**Comment by dwhswenson** [comment]
I'm not sure, but I suspect that Cython can at least help by speeding up function calls and loops (both of which it excels at with very little effort). I'd need to look more into allocation to answer more (and that I will do when I'm not handling this discussion via phone app)

I have the sense that you're less eager to learn and work with Cython than I am. But of course, I can't just do this without your help, since you know all the ins and outs of netcdfplus. This sounds like a long Skype call to be plan how it might be done, followed by several long Skype calls along the way to make sure I don't break things while cythonizing.

You have a PR on the way to solve the issue discussed above, I think? Have that close this issue, and Cythonization of storage can be a separate issue.

**Comment by jhprinz** [comment]
I think that is a good strategy. I wanted to learn cython at some point, but I am a little tired working so much on storage that other things are more interesting. 

Maybe in the meantime we can add some options to AnalysisStorage to tweak the size of the cache (depending on memory) and what to actually include in the caching. Or have specific function for certain parts, so if you want the rate quick do not load the moves, etc...

**Comment by jchodera** [comment]
In addition to spending up NetCDF storage with cython post 1.x, I wonder if
there are high performance distributed Python object storage databases that
might be an alternative (or could be used together with) the NetCDF storage.

**Comment by jhprinz** [comment]
Yes, I thought about this at some point, but could not find something useful for our needs, or let's say it took a lot of iterations to really figure out how we want to store things in what level of flexibility. Just storing it in a database is not too difficult. It would mean to replace a few save functions. 

Even better (and I think that is what you have in mind) would be a better/faster way to pickle the OPS objects. So far I have not found something that can keep track of connections as we do, but I bet this also exists in some way. Just pickling will have a lot of redundancy. The distributed storage approaches can do that since they also address objects with a unique identifier.

I guess that would take more time figuring out. It is very much using NoSQL databases I think. That could be a good approach (I hope not all storage work was pointless though :)

**Comment by jchodera** [comment]
I think there are distributed object store databases that are optimized for this sort of thing. We can look into it over the next few months...

**Comment by jhprinz** [comment]
> We can look into it over the next few months...

That would be good. I might also need this for the adaptive sampling project.

**Comment by jchodera** [comment]
Something like this is in the direction we're looking for:

http://www.zodb.org/en/latest/

And it may be optimal to combine both NetCDF (for snapshots) and an object
store (for Python objects) to get the best of both worlds.

**Comment by jhprinz** [comment]
For reference. I used mongodb for the adaptivemd project. With griefs it can also hold arbitrary large objects. Alas, this will become feasible once we switch from single snapshot to trajectory chunks / snapshot blocks. Otherwise the micromanagement of each snapshot kills performance. I guess we will have a big discussion about this some time in the future.

**Comment by jhprinz** [comment]
Anyway, is this problem solved with the latest PR s? It should now be much faster to read.

**Comment by dwhswenson** [comment]
Yes, I'd say that the biggest problem was solved in #670; enough to close this issue for now (I think we'll revisit file performance later -- faster loading will always be better!)
